### PR TITLE
WIP Add Swift 3 runtime with KituraNet and Dispatch

### PR DESCRIPTION
**Pull request by paulcastro**
This is a WIP to start commentary on the changes.  This adds a Swift 3 container based on the 4-24 snapshot and adds KituraNet.  This also includes an update the CLI so you can create a Swift 3 action using "--type" as follows "wsk action create/update --type swift:3.0 myAction action.swift" .  Type value isn't really checked now since it only applies to Swift 3.

### CODE REVIEW

File: tools/cli/wskaction.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def __init__(self):
 
     def getItemSpecificCommands(self, parser, props):
         subcmd = parser.add_parser('create', help='create new action')
+        subcmd.add_argument('--type', help='the type of version of the runtime')
```
Comment (rabbah) [inline tools/cli/wskaction.py]: type should accept a parameter which indicated the `type` - the CLI could then validate lightly or defer to the controller.

File: tools/cli/wskaction.py [hunk 2]
```diff
@@ -170,7 +174,12 @@ def getExec(self, args, props):
             exe = self.getActionExec(args, props, pipeAction)
         elif args.artifact is not None and os.path.isfile(args.artifact):
             contents = open(args.artifact, 'rb').read()
-            if args.artifact.endswith('.swift'):
+            # FIXME:
```
Comment (rabbah) [inline tools/cli/wskaction.py]: type should not assume `swift3` but rather accept an additional value to be used as the type.

File: tools/ubuntu-setup/ansible.sh [hunk 3]
```diff
@@ -1,7 +1,7 @@
 sudo apt-get install -y software-properties-common
 sudo apt-add-repository -y ppa:ansible/ansible
 sudo apt-get update -y
-sudo apt-get install -y ansible=2.0.1.0-1ppa~trusty
+sudo apt-get install -y ansible=2.0.2.0-1ppa~trusty
```
Comment (rabbah) [inline tools/ubuntu-setup/ansible.sh]: I think this was reverted - rebase master.

File: tools/vagrant/Vagrantfile [hunk 4]
```diff
@@ -6,7 +6,7 @@ BOX = "ubuntu/trusty64-2"
 BOX_URL =  "https://cloud-images.ubuntu.com/vagrant/trusty/current/trusty-server-cloudimg-amd64-vagrant-disk1.box"
 BOX_MEMORY = ENV['BOX_MEMORY'] || '4096'
 BOX_CPUS = ENV['BOX_CPUS'] || '8'
-MACHINE_IP = ENV['MACHINE_IP'] || '192.168.33.13'
+MACHINE_IP = ENV['MACHINE_IP'] || '192.168.56.111'
```
Comment (rabbah) [inline tools/vagrant/Vagrantfile]: ?

File: catalog/samples/helloSwift3.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (rabbah) [inline catalog/samples/helloSwift3.swift]: Can we call this something else other than `helloSwift3`? I don't have a good suggestion yet, alas.

**Comment by rabbah** [comment]
Missing a CLI test to run with `--type swift3`?

File: core/swiftAction/Dockerfile.hold2 [hunk 6]
```diff
@@ -1,27 +0,0 @@
-FROM ubuntu:14.04
```
Comment (psuter) [inline core/swiftAction/Dockerfile.hold2]: If this file gets created then deleted, you should probably squash some commits to make these changes disappear from the history.

File: core/swift3Action/epilogue.swift [hunk 7]
```diff
@@ -1,138 +1,53 @@
...
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+* Copyright 2015-2016 IBM Corporation
+    *
+    * Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (psuter) [inline core/swift3Action/epilogue.swift]: Something weird with whitespace here.

File: common/scala/src/whisk/core/entity/Exec.scala [hunk 8]
```diff
@@ -114,6 +124,7 @@ protected[core] object Exec
             case BlackBoxExec(image)          => JsObject("kind" -> JsString(Exec.BLACKBOX), "image" -> JsString(image))
             case SwiftExec(code)              => JsObject("kind" -> JsString(Exec.SWIFT), "code" -> JsString(code))
             case JavaExec(jar, main)          => JsObject("kind" -> JsString(Exec.JAVA), "jar" -> JsString(jar), "main" -> JsString(main))
+            case Swift3Exec(code)              => JsObject("kind" -> JsString(Exec.SWIFT3), "code" -> JsString(code))
```
Comment (rabbah) [inline common/scala/src/whisk/core/entity/Exec.scala]: @psuter here too - seems to misalign the `=>`

File: common/scala/src/whisk/core/entity/Exec.scala [hunk 9]
```diff
@@ -101,10 +109,12 @@ protected[core] object Exec
     protected[core] val BLACKBOX = "blackbox"
     protected[core] val SWIFT    = "swift"
     protected[core] val JAVA     = "java"
+    protected[core] val SWIFT3 = "swift3"
```
Comment (rabbah) [inline common/scala/src/whisk/core/entity/Exec.scala]: @psuter here too - seems to misalign the `=`

File: core/swift3Action/epilogue.swift [hunk 10]
```diff
@@ -1,138 +1,53 @@
...
+*/
 
 // Deliberate whitespaces above.
 
 /* This code is appended to user-supplied action code.
-   It reads from the standard input, deserializes into JSON and invokes the
-   main function. Currently, actions print strings to stdout. This can evolve once
-   JSON serialization is available in Foundation. */
+ It reads from the standard input, deserializes into JSON and invokes the
+ main function. Currently, actions print strings to stdout. This can evolve once
```
Comment (psuter) [inline core/swift3Action/epilogue.swift]: I think this comment is outdated? Actions don't print to stdout. To be fair, it's outdated in `swiftAction` too.

File: core/swift3Action/epilogue.swift [hunk 11]
```diff
@@ -1,138 +1,53 @@
...
     let inputStr: String = env["WHISK_INPUT"] ?? "{}"
-
-    if let parsed = _whisk_json2dict(inputStr) {
-        let result = main(parsed)
+    
+    if let parsed = _whisk_json2dict(txt: inputStr) {
+        let result = main(args:parsed)
         do {
-            try print(_Whisk_JSONSerialization.serialize(result))
+            let resp = try NSJSONSerialization.data(withJSONObject: result.bridge(), options: [])
```
Comment (psuter) [inline core/swift3Action/epilogue.swift]: Glad this works now!

File: core/swift3Action/proxy.py [hunk 12]
```diff
@@ -29,7 +29,7 @@
 
 SRC_EPILOGUE_FILE = "./epilogue.swift"
 DEST_SCRIPT_FILE = "/swift3Action/spm-build/main.swift"
-DEST_SCRIPT_DIR = "/swif3Action/spm-build"
+DEST_SCRIPT_DIR = "/swift3Action/spm-build"
```
Comment (psuter) [inline core/swift3Action/proxy.py]: another example of a change that should be eliminated when squashing commits

File: common/scala/src/whisk/core/entity/Exec.scala [hunk 13]
```diff
@@ -40,8 +40,8 @@ import java.util.Base64
  * For Java actions, a base64-encoded string representing a jar file is
  * required, as well as the name of the entrypoint class.
  *
- * exec: { kind  : one of "nodejs", "blackbox", "swift",
- *         code  : code to execute if kind is "nodejs" or "swift",
+ * exec: { kind  : one of "nodejs", "blackbox", "swift", "swift3"
```
Comment (psuter) [inline common/scala/src/whisk/core/entity/Exec.scala]: `java` should be listed too. Again, my previous omission, not yours.

File: common/scala/src/whisk/core/entity/Exec.scala [hunk 14]
```diff
@@ -165,6 +185,16 @@ protected[core] object Exec
                     JavaExec(jar, main)
 
                 case _ => throw new DeserializationException(s"'kind' must be one of {${Exec.NODEJS},${Exec.BLACKBOX},${Exec.SWIFT},${Exec.JAVA}}")
+=======
+                case "swift3" =>
+                    val code: String = obj.getFields("code") match {
+                        case Seq(JsString(c)) => c
+                        case _                => throw new DeserializationException(s"'code' must be a string defined in 'exec' for '${Exec.SWIFT}, ${Exec.SWIFT3}' actions")
```
Comment (psuter) [inline common/scala/src/whisk/core/entity/Exec.scala]: Why mention the `swift` kind in the error message?

File: tools/cli/wskaction.py (hunk 1 above)
Comment (psuter) [inline tools/cli/wskaction.py]: Any reason why it's called `type` as opposed to the already-used-internally `kind`?

Comment (sjfink) [inline tools/cli/wskaction.py]: No reason.

File: tests/src/actionContainers/Swift3ActionContainerTests.scala [hunk 15]
```diff
@@ -0,0 +1,148 @@
+/*
```
Comment (psuter) [inline tests/src/actionContainers/Swift3ActionContainerTests.scala]: If you run exactly the same tests as for `whisk/swiftaction`, maybe factor them out? Alternatively: add tests that exercise the swift3 features?

**Comment by psuter** [comment]
Overall it looks sensible to me. Rebasing will make it dramatically easier to review the next iteration (no more swiftAction being modified then reverted etc.). As Rodric points out, we need CLI tests, and we need tests somewhere that exercise the new features (either as CLI tests or as container tests). Finally I'm personally in favor of calling the CLI flag `--kind` since that's already the name of the field in the JSON object.

**Comment by sjfink** [comment]
+1 on --kind

**Comment by paulcastro** [comment]
Kind it is.

File: tools/vagrant/Vagrantfile (hunk 4 above)
Comment (csantanapr) [inline tools/vagrant/Vagrantfile]: @paulcastro @psuter  this IP Address `192.168.33.13` is reference in README.md, if there is no good reason I would say do not change it, as folks might already have this IP scripted in their workflow

https://github.com/openwhisk/openwhisk/blob/437f2ef1288839393b9b026f13b6665edcbd3bd2/README.md#L287
https://github.com/openwhisk/openwhisk/blob/437f2ef1288839393b9b026f13b6665edcbd3bd2/README.md#L290

File: catalog/samples/helloSwift3.swift (hunk 5 above)
Comment (csantanapr) [inline catalog/samples/helloSwift3.swift]: What about catalog/samples/swift3/hello.swift with name "Swift3 : Hello World" ?

Comment (rabbah) [inline catalog/samples/helloSwift3.swift]: Apropos the discussion for #188 and #302 I would opt for a more descriptive name. There are two parts though: what the file is called and what the action is called (they need not be the same). But ideally both are descriptive. This is a minor nit for now but will come to the fore eventually.

Comment (sjfink) [inline catalog/samples/helloSwift3.swift]: "httpGet"?

File: tests/src/actionContainers/Swift3ActionContainerTests.scala (hunk 15 above)
Comment (paulcastro) [inline tests/src/actionContainers/Swift3ActionContainerTests.scala]: I added CLI test for wsk action create --kind, also added a KituraNet/Dispatch test.  Swift3 should also repeat swift tests.

**Comment by paulcastro** [comment]
Superceded by #329 .

**Pull request closed without merging**